### PR TITLE
Publication date filter

### LINKED ISSUE #74: Search by publication date
- [X] Design
- [ ] Implement

### PR DESCRIPTION
**Pull request by ondrae**
This PR lets users filter by SORN publication year. The starting year are both inclusive.

Screen Shot 2020-12-18 at 4 13 45 PM

It has validation and tests included.

Screen Shot 2020-12-18 at 3 41 31 PM

Screen Shot 2020-12-21 at 3 34 47 PM

When accepted, please squash and merge.

Closes #74

### CODE REVIEW

**Comment by peterrowland** [comment]
My intuition seeing 'Ending Year' is that it will include that year in the search, instead of the stop date as it is here. If I wanted my search to include everything up to the most current SORNs, it's awkward to have to enter a date in the future to get the most recent SORNs.

The description text is being used to communicate the format to be entered, but i think we could do that with a hint like '_YYYY_' in the field when empty. Instead I think we should use that space and say:

**Starting Year**
SORNs published starting from January 1st of the year in this field.

**Ending Year**
SORNs published up to December 31st of the year in this field.

**Comment by igorkorenfeld** [comment]
I agree, I think that most users would expect Ending Year to be inclusive as well, and it may be confusing that one is inclusive and one is not.

I'd be interested to test out that explainer text.

**Comment by ondrae** [comment]
Peter pointed out we need validation that starting year is before ending year.

**Comment by ondrae** [comment]
Okay, learning how to do the validation on the frontend:

- https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation
- http://jsfiddle.net/ryeballar/qbL53/4/
- https://stackoverflow.com/questions/22702629/how-to-create-html5-custom-validation

**Comment by ondrae** [comment]
Okay, added a custom validation and made the ending year inclusive.

Screen Shot 2020-12-21 at 3 34 47 PM

File: app/javascript/packs/search.js [hunk 1]
```diff
@@ -39,7 +39,9 @@ function checkboxesFromUrl(elementName) {
 }
 
 function publicationDateValidation(){
-  if ($("#starting_year").val() > $("#ending_year").val()) {
+  startYear = parseInt($("#starting_year").val())
+  endYear = parseInt($("#ending_year").val())
+  if (startYear > endYear) {
     $("#starting_year")[0].setCustomValidity("Starting year should be lower than the ending year.");
```
Comment (igorkorenfeld) [inline app/javascript/packs/search.js:45]: I think "earlier" or "before" might be more clear than "lower" since we are talking about dates.

**Comment by igorkorenfeld** [comment]
Very cool!

I had a couple of issues:

1. When I tried to test it out locally neither the character count nor the start-end date validations poped up for me. 

2. When I go to any page of the results other than the first page, the agencies filter is opened, and neither filter will close on click. Maybe our javascript only loads on the first page? 

Also added in the in-line comment about the validation message wording.

**Comment by ondrae** [comment]
Igor, it sounds like javascript isn't working at all for you locally. Let's pair and look at it.

**Comment by ondrae** [comment]
Igor's js issues are separate and part of #69

File: app/javascript/packs/search.js [hunk 2]
```diff
@@ -1,4 +1,5 @@
-$( function () {
+// Needs to wait for turbolink to load, for the system specs to work
```
Comment (ondrae) [inline app/javascript/packs/search.js:1]: Will need to get rid of this turbolinks code if we merge in #91

File: spec/requests/search_request_spec.rb [hunk 3]
```diff
@@ -131,10 +127,36 @@
...
+
+    it "only returns the older sorn in date range" do
+      get "/search?ending_year=2001"
+
+      expect(response.body).to include "2000-01-13" # Older sorn date
+      expect(response.body).not_to include "NEW SORN"
+      expect(response.body).not_to include "2019-01-13" # Newer sorn date
+    end
+
+    it "ending year is inclusive" do
```
Comment (lauraGgit) [inline spec/requests/search_request_spec.rb:155]: i could go either way, but would it make sense to have a test case to just check that the 2 search params can be on the `/search` endpoint?

**Review by lauraGgit — COMMENTED** [review]
looks good! couple of minor comments that i don't think are totally necessary but could be nice additions

File: app/javascript/packs/search.js [hunk 4]
```diff
@@ -31,4 +35,14 @@ function checkboxesFromUrl(elementName) {
       $(`#${elementName}-${data}`).prop("checked", true)
     });
   }
+}
+
+function publicationDateValidation(){
+  startYear = parseInt($("#starting_year").val())
```
[the remark this answers is not in the document]
Comment (ondrae) [reply app/javascript/packs/search.js:41]: @igorkorenfeld what would a good validation message be for this?
> "We have SORNs from 1994 til today."

We also have 1994 in the hint above the field.

[the remark this answers is not in the document]
Comment (ondrae) [reply app/javascript/packs/search.js:41]: > 94 til infinity

[the remark this answers is not in the document]
Comment (igorkorenfeld) [reply app/javascript/packs/search.js:41]: We should do a content/tone audit later, but maybe let's start with something like "Sorry, this tool only contains SORNs starting from 1994. Please enter a later starting year" ?

[the remark this answers is not in the document]
Comment (ondrae) [reply app/javascript/packs/search.js:41]: Okay done.

File: app/views/sorns/_publication_date.html.erb [hunk 5]
```diff
@@ -7,7 +7,7 @@
 
   <div class="usa-form-group">
     <label class="usa-label" for="ending_year">Ending year</label>
-    <span class="usa-hint">For example: 2020</span>
+    <span class="usa-hint">For example: <%= Date.today.year %></span>
```
Comment (igorkorenfeld) [inline app/views/sorns/_publication_date.html.erb:10]: This is cool!

**Comment by igorkorenfeld** [comment]
I am still waiting on my local db to repopulate with the older SORNs but testing the before/after and pre 1994 validation worked great. Will update when it finishes pull SORNs, but if it takes too long toady, maybe @peterrowland can look over it next week?

**Review by peterrowland — APPROVED** [review]
approved!

**Pull request merged**